HadesScheduler/k8s/job.go

```diff
@@ -0,0 +1,150 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/ls1intum/hades/shared/payload"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+type K8sJob struct {
+	payload.QueuePayload
+	k8sClient        *kubernetes.Clientset
+	namespace        string
+	sharedVolumeName string
+}
+
+// Schedules a Hades Job on the Kubernetes cluster
+func (k8sJob K8sJob) execute(ctx context.Context) error {
+	slog.Info("Scheduling job", "id", k8sJob.ID)
+
+	slog.Debug("Create buildscript ConfigMap")
+	configMap := k8sJob.configMapSpec()
+	slog.Debug("ConfigMap spec", "config_map", configMap)
+
+	slog.Debug("Apply buildscirpt ConfigMap to Kubernetes")
+	cm, err := k8sJob.k8sClient.CoreV1().ConfigMaps(k8sJob.namespace).Create(ctx, configMap, metav1.CreateOptions{})
+	if err != nil {
+		slog.With("error", err).Error("Failed to create ConfigMap")
+		return err
+	}
+	slog.Info("Successfully created buildscirpt ConfigMap", "name", cm.Name)
+
+	slog.Debug("Assembling PodSpec")
+	jobPodSpec := corev1.Pod{
+
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      k8sJob.ID.String(),
+			Namespace: k8sJob.namespace,
+		},
+
+		Spec: corev1.PodSpec{
+			InitContainers: k8sJob.containerSpec(), // each step is represented by a init container to ensure that the build script is executed in the correct order
+			Containers: []corev1.Container{ // dummy container to signal the end of the build - this is a temporary solution as we need to have a container defined in the pod spec
+				{
+					Name:    "dummy",
+					Image:   "busybox",
+					Command: []string{"sh", "-c", "echo 'build completed'"},
+				},
+			},
+			Volumes:       k8sJob.volumeSpec(*configMap),
+			RestartPolicy: corev1.RestartPolicyNever,
+		},
+	}
+	slog.Debug("PodSpec", "spec", jobPodSpec)
+
+	slog.Info("Apply PodSpec to Kubernetes")
+	_, err = k8sJob.k8sClient.CoreV1().Pods(k8sJob.namespace).Create(ctx, &jobPodSpec, metav1.CreateOptions{})
+	if err != nil {
+		slog.With("error", err).Error("Failed to create Pod")
+		return err
+	}
+
+	return nil
+}
+
+// Creates a ConfigMapSpec containing the build script of each step of the job
+// Based on this configMap, individual step volumes are created and mounted inside the respective container
+func (k8sJob K8sJob) configMapSpec() *corev1.ConfigMap {
+	configMap := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      k8sJob.ID.String(),
+			Namespace: k8sJob.namespace,
+		},
+		Data: map[string]string{},
```

Comment:
Do we need to specify it? Or is the default always am empty map?

Reply:
Yes - We put in the steps later which require the map to be initialized